src/main/java/de/hysky/skyblocker/skyblock/dungeon/puzzle/ThreeWeirdos.java

```diff
@@ -1,39 +1,85 @@
 import net.minecraft.entity.decoration.ArmorStandEntity;
 import net.minecraft.text.Text;
+import net.minecraft.util.ActionResult;
 import net.minecraft.util.Formatting;
+import net.minecraft.util.hit.HitResult;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Box;
+import net.minecraft.world.World;
 
+import java.util.List;
 import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
-public class ThreeWeirdos extends ChatPatternListener {
-    public ThreeWeirdos() {
-        super("^\\[NPC] ([A-Z][a-z]+): (?:The reward is(?: not in my chest!|n't in any of our chests\\.)|My chest (?:doesn't have the reward\\. We are all telling the truth\\.|has the reward and I'm telling the truth!)|At least one of them is lying, and the reward is not in [A-Z][a-z]+'s chest\\!|Both of them are telling the truth\\. Also, [A-Z][a-z]+ has the reward in their chest\\!)$");
+public class ThreeWeirdos extends DungeonPuzzle {
+    protected static final Pattern PATTERN = Pattern.compile("^\\[NPC] ([A-Z][a-z]+): (?:The reward is(?: not in my chest!|n't in any of our chests\\.)|My chest (?:doesn't have the reward\\. We are all telling the truth\\.|has the reward and I'm telling the truth!)|At least one of them is lying, and the reward is not in [A-Z][a-z]+'s chest!|Both of them are telling the truth\\. Also, [A-Z][a-z]+ has the reward in their chest!)$");
+    private static final float[] GREEN_COLOR_COMPONENTS = new float[]{0, 1, 0};
+    private static BlockPos pos;
+
+    private ThreeWeirdos() {
+        super("three-weirdos", "three-chests");
+        ClientReceiveMessageEvents.GAME.register((message, overlay) -> {
+            World world = MinecraftClient.getInstance().world;
+            if (overlay || !shouldSolve() || !SkyblockerConfigManager.get().dungeons.puzzleSolvers.solveThreeWeirdos || world == null || !DungeonManager.isCurrentRoomMatched()) return;
+
+            @SuppressWarnings("DataFlowIssue")
+            Matcher matcher = PATTERN.matcher(Formatting.strip(message.getString()));
+            if (!matcher.matches()) return;
+            String name = matcher.group(1);
+            Room room = DungeonManager.getCurrentRoom();
+
+            checkForNPC(world, room, new BlockPos(13, 69, 24), name);
+            checkForNPC(world, room, new BlockPos(15, 69, 25), name);
+            checkForNPC(world, room, new BlockPos(17, 69, 24), name);
+        });
+        UseBlockCallback.EVENT.register((player, world, hand, blockHitResult) -> {
+            if (blockHitResult.getType() == HitResult.Type.BLOCK && blockHitResult.getBlockPos().equals(pos)) {
+                pos = null;
+            }
+            return ActionResult.PASS;
+        });
     }
 
-    @Override
-    public ChatFilterResult state() {
-        return SkyblockerConfigManager.get().dungeons.puzzleSolvers.solveThreeWeirdos ? null : ChatFilterResult.PASS;
+    public static void init() {
+        new ThreeWeirdos();
```

Comment:
Its probably more consistent to put this in an INSTANCE variable like the other puzzle solvers

Reply:
Currently the others only have the instance saved when the instance is needed to be called. Don’t think I will bother with this.